lib/auth/auth_with_roles_test.go

```diff
@@ -2812,53 +2813,91 @@ func TestListResources_SearchAsRoles(t *testing.T) {
-	require.Len(t, resp.Resources, 1)
-	require.Equal(t, resp.Resources[0].GetName(), testNodes[0].GetName())
+	for _, tc := range []struct {
+		desc                   string
+		requestOpt             func(*proto.ListResourcesRequest)
+		expectNodes            []string
+		expectSearchEventRoles []string
+	}{
+		{
+			desc:        "basic",
+			expectNodes: []string{testNodes[0].GetName()},
+		},
+		{
+			desc: "search as roles",
+			requestOpt: func(req *proto.ListResourcesRequest) {
+				req.UseSearchAsRoles = true
+			},
+			expectNodes:            []string{testNodes[0].GetName(), testNodes[1].GetName()},
+			expectSearchEventRoles: []string{role.GetName(), searchAsRole.GetName()},
+		},
+		{
+			desc: "preview as roles",
+			requestOpt: func(req *proto.ListResourcesRequest) {
+				req.UsePreviewAsRoles = true
+			},
+			expectNodes:            []string{testNodes[0].GetName(), testNodes[2].GetName()},
+			expectSearchEventRoles: []string{role.GetName(), previewAsRole.GetName()},
+		},
+		{
+			desc: "both",
+			requestOpt: func(req *proto.ListResourcesRequest) {
+				req.UseSearchAsRoles = true
+				req.UsePreviewAsRoles = true
+			},
+			expectNodes:            []string{testNodes[0].GetName(), testNodes[1].GetName(), testNodes[2].GetName()},
+			expectSearchEventRoles: []string{role.GetName(), searchAsRole.GetName(), previewAsRole.GetName()},
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			req := proto.ListResourcesRequest{
+				ResourceType: types.KindNode,
+				Limit:        int32(len(testNodes)),
+			}
+			if tc.requestOpt != nil {
+				tc.requestOpt(&req)
+			}
+			resp, err := clt.ListResources(ctx, req)
+			require.NoError(t, err)
+			require.Len(t, resp.Resources, len(tc.expectNodes))
+			var gotNodes []string
+			for _, node := range resp.Resources {
+				gotNodes = append(gotNodes, node.GetName())
+			}
+			require.ElementsMatch(t, tc.expectNodes, gotNodes)
 
-	// ListResources with UseSearchAsRoles returns both nodes
-	resp, err = clt.ListResources(ctx, proto.ListResourcesRequest{
-		ResourceType:     types.KindNode,
-		Limit:            int32(len(testNodes)),
-		UseSearchAsRoles: true,
-	})
-	require.NoError(t, err)
-	require.Len(t, resp.Resources, len(testNodes))
-	var expectedNodes []string
-	for _, node := range testNodes {
-		expectedNodes = append(expectedNodes, node.GetName())
-	}
-	var gotNodes []string
-	for _, node := range resp.Resources {
-		gotNodes = append(gotNodes, node.GetName())
+			if len(tc.expectSearchEventRoles) > 0 {
+				require.Eventually(t, func() bool {
+					// make sure an audit event is logged for the search
+					auditEvents, _, err := srv.AuthServer.AuditLog.SearchEvents(time.Time{}, time.Now(), "", []string{events.AccessRequestResourceSearch}, 10, 0, "")
+					require.NoError(t, err)
+					for _, event := range auditEvents {
+						if searchEvent, ok := event.(*apievents.AccessRequestResourceSearch); ok {
+							if reflect.DeepEqual(tc.expectSearchEventRoles, searchEvent.SearchAsRoles) {
```

Comment:
Can we use `cmp.Diff` instead of `reflect.Equal` ? I tend to find it makes it easier to debug a misbehaving test.